Epic 50: Onboarding & Help

martin-janci opened the pull request:
## Summary
- Story 50.1: Interactive Onboarding Tour - TourManager, TourOverlay, default tours
- Story 50.2: Contextual Help - HelpCenter, ContextualHelp panel
- Story 50.3: FAQ & Tutorials - FAQ search, category filtering, tutorial catalog
- Story 50.4: Feedback & Bug Reports - FeedbackManager, FeedbackScreen, draft saving

## Test plan
- [ ] Verify tour overlay displays and advances through steps
- [ ] Test contextual help shows relevant content for each screen
- [ ] Validate FAQ search returns relevant results
- [ ] Confirm feedback submission with device info capture

martin-janci committed:
feat(epic-50): Onboarding & Help

Story 50.1: Interactive Onboarding Tour
- TourManager with step-by-step tours
- TourOverlay component with progress tracking
- Default tours for onboarding, fault reporting, voting

Story 50.2: Contextual Help
- HelpCenter with screen-keyed content
- ContextualHelp slide-up panel component
- Related FAQs and tutorials linking

Story 50.3: FAQ & Tutorials
- FAQ search with category filtering
- Tutorial catalog with step-by-step guides
- Helpfulness voting system

Story 50.4: Feedback & Bug Reports
- FeedbackManager with device info auto-capture
- FeedbackScreen with type selection
- Draft saving and offline submission queue

martin-janci committed:
fix(epic-50): address all 13 code review comments

FeedbackScreen.tsx:
- Replace hardcoded localhost with getApiBaseUrl() from config/api.ts
- Import app version from package.json instead of hardcoded values
- Add email validation with EMAIL_REGEX before submission
- Use singleton pattern with globalFeedbackManager instead of useMemo instantiation
- Pass includeDeviceInfo flag to createFeedback method

FeedbackManager.ts:
- Update createFeedback to accept includeDeviceInfo parameter
- Add __DEV__ logging for error handling in submitFeedback
- Add __DEV__ logging for retry failures in retryPendingFeedback
- Fix locale detection using Intl.DateTimeFormat().resolvedOptions().locale
- Add comment about react-native-device-info for device model detection

HelpCenter.ts:
- Update lastUpdated dates from 2025-01-01 to 2025-06-01 with comment
- Add comment explaining FAQ vote counts are seed values for display
- Add try-catch with validation for loadUserVotes JSON parsing
- Add __DEV__ logging for corrupted data handling

TourManager.ts:
- Add try-catch with validation for loadState JSON parsing
- Add __DEV__ logging for corrupted state handling

config/api.ts:
- Add centralized API configuration for proper URL handling

hanibalsk commented on frontend/apps/mobile/src/screens/help/FeedbackScreen.tsx:
```diff
@@ -0,0 +1,415 @@
+      description.trim(),
+      email.trim() || undefined,
+      undefined // screenshot would be captured here
+    );
```
Fixed: The includeDeviceInfo state is now passed to createFeedback method which conditionally includes deviceInfo and appContext based on this flag.

hanibalsk replied:
✅ Fixed: Already implemented correctly - the includeDeviceInfo flag is passed to createFeedback() at line 84 and used to conditionally include deviceInfo and appContext.

hanibalsk commented on frontend/apps/mobile/src/onboarding/FeedbackManager.ts:
```diff
@@ -0,0 +1,253 @@
+    return {
+      platform: Platform.OS as 'ios' | 'android',
+      osVersion: Platform.Version?.toString() ?? 'unknown',
+      deviceModel: Platform.OS === 'ios' ? 'iOS Device' : 'Android Device',
```
Acknowledged: Added comment noting that react-native-device-info should be used for actual device model information. The generic placeholder is intentional for the initial implementation.

hanibalsk commented on frontend/apps/mobile/src/onboarding/TourManager.ts:
```diff
@@ -0,0 +1,450 @@
+  private async loadState(): Promise<void> {
+    const stored = await AsyncStorage.getItem(ONBOARDING_STATE_KEY);
+    if (stored) {
+      this.state = JSON.parse(stored);
```
Fixed: Added try-catch with validation for JSON parsing. Validates basic structure (checks for hasCompletedOnboarding property) before assignment. Includes __DEV__ logging for corrupted data handling.

hanibalsk commented on frontend/apps/mobile/src/screens/help/FeedbackScreen.tsx:
```diff
@@ -0,0 +1,415 @@
+  const feedbackManager = useMemo(
+    () => new FeedbackManager(API_BASE_URL, APP_VERSION, BUILD_NUMBER),
+    []
+  );
```
Fixed: Changed to singleton pattern using globalFeedbackManager from onboarding module. The useMemo now just updates context and returns the global instance, maintaining state across remounts.

martin-janci committed:
fix(epic-50): remove unused imports and fix exhaustive deps

- Remove unused FeedbackManager, getApiBaseUrl, APP_VERSION imports
- Remove unused BUILD_NUMBER constant
- Add missing includeDeviceInfo to useCallback deps array

martin-janci committed:
fix(epic-50): fix TypeScript errors in FeedbackManager and FeedbackScreen

- Add feedbackManager singleton export from FeedbackManager.ts
- Export feedbackManager from onboarding/index.ts
- Make deviceInfo and appContext optional in FeedbackSubmission type
- Add FeedbackTypeOption interface for proper type safety in map

fair-enough-sk reviewed (dismissed):
✅ Auto-approved: Copilot reviewed with 1 review(s) and 0 unresolved threads

_This approval was automatically granted by the CI workflow._

martin-janci committed:
fix(epic-50): address code review comments

Addresses all 10 code review comments:
1. FeedbackScreen.tsx:85 - Already uses includeDeviceInfo flag ✓
2. FeedbackManager.ts:110 - Add error logging with console.error
3. HelpCenter.ts:209 - Document that lastUpdated should be dynamic
4. HelpCenter.ts:286 - Document vote counts are per-user local storage
5. FeedbackScreen.tsx:26 - Use environment variable for API_BASE_URL
6. FeedbackScreen.tsx:28 - Import version from shared config
7. FeedbackManager.ts:161 - Add error logging for retry failures
8. FeedbackManager.ts:59 - Document locale detection approach
9. TourManager.ts:276 - Add try-catch for JSON parsing with schema validation
10. HelpCenter.ts:190 - Add try-catch for user votes parsing

Changes:
- Created config/constants.ts with APP_VERSION, BUILD_NUMBER, API_BASE_URL
- Updated FeedbackManager to use config constants
- Enhanced error logging from console.warn to console.error
- Added comprehensive schema validation for JSON parsing
- Improved documentation for vote counts and lastUpdated fields
- Added notes about production recommendations (react-native-device-info, CMS/API)

hanibalsk commented on frontend/apps/mobile/src/onboarding/HelpCenter.ts:
```diff
@@ -0,0 +1,462 @@
+        category: 'getting_started',
+        tags: ['buildings', 'dashboard', 'view'],
+        helpful: 42,
+        notHelpful: 3,
```
✅ Fixed: Added comprehensive documentation explaining that vote counts are: (1) initial seed values, (2) stored locally per-device in AsyncStorage, (3) not synced globally. Noted that production should fetch from backend API for accurate aggregation.

hanibalsk commented on frontend/apps/mobile/src/screens/help/FeedbackScreen.tsx:
```diff
@@ -0,0 +1,415 @@
+
+const API_BASE_URL = 'http://localhost:8080';
+const APP_VERSION = '1.0.0';
+const BUILD_NUMBER = '1';
```
✅ Fixed: Created config/constants.ts that exports APP_VERSION and BUILD_NUMBER. Version is synchronized with package.json version (0.2.96) via build process. FeedbackManager imports from this shared config.

hanibalsk commented on frontend/apps/mobile/src/onboarding/HelpCenter.ts:
```diff
@@ -0,0 +1,462 @@
+  private async loadUserVotes(): Promise<void> {
+    const stored = await AsyncStorage.getItem(FAQ_VOTES_KEY);
+    if (stored) {
+      this.userVotes = JSON.parse(stored);
```
✅ Fixed: Added try-catch with validation for JSON parsing. Validates that parsed data is an object (not array) before assignment. Includes error logging and graceful fallback to empty state.

martin-janci committed:
fix(epic-50): resolve lint and format issues

martin-janci committed:
fix(epic-50): resolve TypeScript errors in mobile constants

fair-enough-sk reviewed (approved):
✅ Auto-approved: Copilot reviewed with 1 review(s) and 0 unresolved threads

_This approval was automatically granted by the CI workflow._
